kernel/src/checkpoint/log_replay.rs

```diff
@@ -0,0 +1,603 @@
+//! The [`CheckpointLogReplayProcessor`] implements specialized log replay logic for creating
+//! checkpoint files. It processes log files in reverse chronological order (newest to oldest)
+//! and selects only the minimal set of actions needed to represent the table state at a given version.
+//!
+//! ## Filtering Process
+//!
+//! For checkpoint creation, this processor applies several filtering and deduplication
+//! steps to each batch of log actions:
+//!
+//! 1. **Protocol and Metadata**: Retains only the latest protocol and metadata actions.
+//! 2. **Transactions**: Keeps the most recent action for each unique transaction (by app ID).
+//! 3. **File Actions**: Deduplicates file actions (add/remove) by path and deletion vector ID,
+//!    keeping only the latest valid action.
+//! 4. **Tombstones**: Excludes expired remove actions older than `minimum_file_retention_timestamp`.
+//!
+//! ## Architecture
+//!
+//! - [`CheckpointVisitor`]: Implements [`RowVisitor`] to examine each action in a batch and
+//!   determine if it should be included in the checkpoint. It maintains state for deduplication
+//!   across multiple actions in a batch and efficiently handles all filtering rules.
+//!
+//! - [`CheckpointLogReplayProcessor`]: Implements the [`LogReplayProcessor`] trait and orchestrates
+//!   the overall process. For each batch of log actions, it:
+//!   1. Creates a visitor with the current deduplication state
+//!   2. Applies the visitor to filter actions in the batch
+//!   3. Updates counters and state for cross-batch deduplication
+//!   4. Produces a [`CheckpointData`] result which includes a selection vector indicating which
+//!      actions should be included in the checkpoint file
+use std::collections::HashSet;
+use std::sync::LazyLock;
+
+use crate::engine_data::{GetData, RowVisitor, TypedGetData as _};
+use crate::log_replay::{FileActionDeduplicator, FileActionKey};
+use crate::schema::{column_name, ColumnName, ColumnNamesAndTypes, DataType};
+use crate::utils::require;
+use crate::{DeltaResult, Error};
+
+/// A visitor that filters actions for inclusion in a V1 spec checkpoint file.
+///
+/// This visitor processes actions in newest-to-oldest order (as they appear in log
+/// replay) and applies deduplication logic for both file and non-file actions to
+/// produce the minimal state representation for the table.
```

Comment:
same nit as above: we produce the actions to include in a checkpoint not minimal state of table